library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/async/RecordedDataQueueHandler.kt

```diff
@@ -0,0 +1,175 @@
+
+        return item
+    }
+
+    /**
+     * Goes through the queue one item at a time for as long as there are items in the queue.
+     * If an item is ready to be consumed, it is processed.
+     * If an invalid item is encountered, it is removed (invalid items are possible
+     * for example if a snapshot failed to traverse the tree).
+     * If neither of the previous conditions occurs, the loop breaks.
+     */
+    @MainThread
+    @Synchronized
+    internal fun tryToConsumeItems() {
+        // no need to create a thread if the queue is empty
+        if (recordedDataQueue.isEmpty()) {
+            return
+        }
+
+        // currentTime needs to be obtained on the uithread
+        val currentTime = timeProvider.getDeviceTimestamp()
+
+        @Suppress("SwallowedException", "TooGenericExceptionCaught")
+        try {
+            executorService.execute {
+                triggerProcessingLoop(currentTime)
+            }
+        } catch (e: RejectedExecutionException) {
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: NullPointerException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        }
+    }
+
+    @WorkerThread
+    private fun triggerProcessingLoop(currentTime: Long) {
+        while (recordedDataQueue.isNotEmpty()) {
+            val nextItem = recordedDataQueue.peek()
+
+            if (nextItem != null) {
+                if (shouldRemoveItem(nextItem, currentTime)) {
+                    recordedDataQueue.poll()
+                } else if (nextItem.isReady()) {
+                    processItem()
+                } else {
+                    break
+                }
+            }
+        }
+    }
+
+    private fun processItem() {
+        when (val item = recordedDataQueue.poll()) {
+            is SnapshotRecordedDataQueueItem ->
+                processSnapshotEvent(item)
+            is TouchEventRecordedDataQueueItem ->
+                processTouchEvent(item)
+        }
+    }
+
+    private fun processSnapshotEvent(item: SnapshotRecordedDataQueueItem) {
+        processor.processScreenSnapshots(
+            nodes = item.nodes,
+            systemInformation = item.systemInformation,
+            newContext = item.newRumContext,
+            prevContext = item.prevRumContext,
+            timestamp = item.timestamp
+        )
+    }
+
+    private fun processTouchEvent(item: TouchEventRecordedDataQueueItem) {
+        processor.processTouchEventsRecords(
+            newContext = item.newRumContext,
+            touchEventsRecords = item.touchData
+        )
+    }
+
+    private fun shouldRemoveItem(recordedDataQueueItem: RecordedDataQueueItem, currentTime: Long) =
+        !recordedDataQueueItem.isValid() || isTooOld(currentTime, recordedDataQueueItem)
+
+    private fun isTooOld(currentTime: Long, recordedDataQueueItem: RecordedDataQueueItem): Boolean =
```

Comment:
I'm wondering if we should ever consider item as "too old"

Should we process all of them regardless the time? What the strategy here?

Reply:
The thing is that if we don't define a threshold for an item to be too old then an item that for whatever reason had a failure resolving it's bitmaps (pendingImages > 0) could get stuck at the head of the queue forever